nifi-nar-bundles/nifi-jolt-record-bundle/nifi-jolt-record-processors/src/main/java/org/apache/nifi/processors/jolt/record/JoltTransformRecord.java

```diff
@@ -0,0 +1,522 @@
+                    }
+                } else {
+                    //for validation we want to be able to ensure the spec is syntactically correct and not try to resolve variables since they may not exist yet
+                    Object specJson = SORTR.getValue().equals(transform) ? null : JsonUtils.jsonToObject(specValue.replaceAll("\\$\\{", "\\\\\\\\\\$\\{"), DEFAULT_CHARSET);
+
+                    if (CUSTOMR.getValue().equals(transform)) {
+                        if (StringUtils.isEmpty(customTransform)) {
+                            final String customMessage = "A custom transformation class should be provided. ";
+                            results.add(new ValidationResult.Builder().valid(false)
+                                    .explanation(customMessage)
+                                    .build());
+                        } else {
+                            TransformFactory.getCustomTransform(customClassLoader, customTransform, specJson);
+                        }
+                    } else {
+                        TransformFactory.getTransform(customClassLoader, transform, specJson);
+                    }
+                }
+            } catch (final Exception e) {
+                getLogger().info("Processor is not valid - " + e.toString());
+                String message = "Specification not valid for the selected transformation.";
+                results.add(new ValidationResult.Builder().valid(false)
+                        .explanation(message)
+                        .build());
+            }
+        }
+
+        return results;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void onTrigger(final ProcessContext context, ProcessSession session) throws ProcessException {
+        final FlowFile original = session.get();
+        if (original == null) {
+            return;
+        }
+
+        final ComponentLog logger = getLogger();
+        final StopWatch stopWatch = new StopWatch(true);
+
+        final String transformStrategy = context.getProperty(TRANSFORM_STRATEGY).getValue();
+        final RecordReaderFactory readerFactory = context.getProperty(RECORD_READER).asControllerService(RecordReaderFactory.class);
+        final RecordSetWriterFactory writerFactory = context.getProperty(RECORD_WRITER).asControllerService(RecordSetWriterFactory.class);
+
+        final RecordSchema schema;
+        List<Map<String, Object>> recordList = new ArrayList<>();
+        try (final InputStream in = session.read(original);
+             final RecordReader reader = readerFactory.createRecordReader(original, in, getLogger())) {
+            schema = writerFactory.getSchema(original.getAttributes(), reader.getSchema());
+            Record record;
+
+            while ((record = reader.nextRecord()) != null) {
+                recordList.add((Map<String, Object>) DataTypeUtils.convertRecordFieldtoObject(record, RecordFieldType.RECORD.getRecordDataType(record.getSchema())));
```

Comment:
If I understand this correctly, we're loading to memory the full content of the flow file. Could be nice to add the ``@SystemResourceConsideration`` annotation to the class to make users aware of it.

Reply:
Agreed, will add it